### DISCUSSION TITLE
Building Plus Toolkit in Ubuntu 22.04.4 LTS

### DISCUSSION THREAD

**Discussion by Karl-Philippe** [PlusToolkit/PlusLib#1167] [Q&A]
Hi,

I am trying to build Plus Toolkit (master branch) on Ubuntu 22.04.04 LTS.

I have followed the Ubuntu 22.04 steps on this page: [Build Instructions for Linux](https://github.com/PlusToolkit/PlusBuild/blob/master/Docs/BuildInstructionsLinux.md)

> sudo apt update && sudo apt install git build-essential cmake \
>     qtbase5-dev qt5-qmake qtmultimedia5-dev qttools5-dev libglvnd-dev libqt5xmlpatterns5-dev \
>     qtbase5-private-dev libqt5x11extras5-dev libxt-dev qtdeclarative5-dev libqt5webenginewidgets5 qml-module-qtquick\*
> 
> git clone https://github.com/PlusToolkit/PlusBuild.git
> mkdir PlusBuild-bin
> cd PlusBuild-bin
> cmake ../PlusBuild
> 
> make

However, while building Plus, I am encountering the following error:

```
[100%] Building CXX object GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/qmlvtkplugin_autogen/mocs_compilation.cpp.o
[100%] Building CXX object GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/QQmlVTKPlugin.cxx.o
[100%] Linking CXX shared module ../../../lib/qml/VTK.9.1/libqmlvtkpluginD.so
Generating qmltypes file using qmlplugindump
QQmlComponent: Component is not ready
make[5]: *** [GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/build.make:150: lib/qml/VTK.9.1/libqmlvtkpluginD.so] Error 3
make[5]: *** Deleting file 'lib/qml/VTK.9.1/libqmlvtkpluginD.so'
make[4]: *** [CMakeFiles/Makefile2:11476: GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/all] Error 2
make[3]: *** [Makefile:136: all] Error 2
make[2]: *** [CMakeFiles/vtk.dir/build.make:92: vtk-prefix/src/vtk-stamp/vtk-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:125: CMakeFiles/vtk.dir/all] Error 2
make: *** [Makefile:101: all] Error 2
```

Has anyone else experienced this issue? Do you have any suggestions?

Thanks,

**Comment by adamrankin** [comment]
Could you go into `PlusBuild-bin/vtk-bin` and run `make` and post the errors from that build? The errors here are just superbuild errors which don't give us great information.

**Reply by Karl-Philippe** [reply]
Yes of course, here are the errors from the `PlusBuild-bin/vtk-bin` build:

```
[ 95%] Automatic MOC for target qmlvtkplugin
[ 95%] Built target qmlvtkplugin_autogen
Consolidate compiler generated dependencies of target qmlvtkplugin
[ 95%] Linking CXX shared module ../../../lib/qml/VTK.9.1/libqmlvtkpluginD.so
Generating qmltypes file using qmlplugindump
QQmlComponent: Component is not ready
make[2]: *** [GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/build.make:150: lib/qml/VTK.9.1/libqmlvtkpluginD.so] Error 3
make[2]: *** Deleting file 'lib/qml/VTK.9.1/libqmlvtkpluginD.so'
make[1]: *** [CMakeFiles/Makefile2:11476: GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Comment by adamrankin** [comment]
This might be it:

From https://gitlab.kitware.com/vtk/vtk/-/issues/18219

> If anyone came across this on Ubuntu 20.04 using Qt 5 from the distro packages and none of the above helped, you are most likely missing these packages. `sudo apt install qml-module-qtquick*` fixed it for me.

**Reply by adamrankin** [reply]
If that's not it, apparently re-running cmake on vtk-bin might work

**Reply by Karl-Philippe** [reply]
It seemed that `qml-module-qtquick*` was already installed on my machine. However, after installing the libraries mentioned in [https://gitlab.kitware.com/vtk/vtk/-/issues/18219]:
> 1. From Addidional Libraries in QT Install, make sure `Qt WebEngine`, `Qt Positioning`, `Qt WebChannel` and `Qt WebView` are installed in Qt.

By running `sudo apt install qtwebengine5-dev qtpositioning5-dev qtwebchannel5-dev qml-module-qtwebview`

And then after clearing the build directory and re-running make on `PlusBuild-bin/vtk-bin`, VTK built without errors:

```
[100%] Built target ITKReview
[100%] Built target itkTestDriver
[100%] Built target ITKReview-all
[100%] Built target ITKTestKernel-all
```

However, there is still a build error when building `PlusBuild-bin` (`make -j16`):

```
Consolidate compiler generated dependencies of target ITKReview
[100%] Built target ITKIOTransformHDF5-all
[100%] Linking CXX shared library /home/kpbeaudet/Documents/PlusBuild-bin/bin/libITKReview-5.3.so
[100%] Built target ITKReview
[100%] Built target itkTestDriver
[100%] Built target ITKReview-all
[100%] Built target ITKTestKernel-all
[ 39%] No install step for 'itk'
[ 40%] Completed 'itk'
[ 43%] Built target itk
make: *** [Makefile:101: all] Error 2
```

**Reply by Karl-Philippe** [reply]
Actually I still have the same error when running `make` in `PlusBuild-bin`:
```
Generating qmltypes file using qmlplugindump
QQmlComponent: Component is not ready
make[5]: *** [GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/build.make:150: lib/qml/VTK.9.1/libqmlvtkpluginD.so] Error 3
make[5]: *** Deleting file 'lib/qml/VTK.9.1/libqmlvtkpluginD.so'
make[4]: *** [CMakeFiles/Makefile2:11494: GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/all] Error 2
make[3]: *** [Makefile:136: all] Error 2
make[2]: *** [CMakeFiles/vtk.dir/build.make:92: vtk-prefix/src/vtk-stamp/vtk-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:125: CMakeFiles/vtk.dir/all] Error 2
make: *** [Makefile:101: all] Error 2
```

**Reply by adamrankin** [reply]
You'll have to identify which target is failing. PlusBuild is a superbuild that builds many dependent projects. Do `make` with no `-j` argument for an easier time identifying the last failed project.

**Comment by Karl-Philippe** [comment]
I tested on multiple machines with Ubuntu 22.04 and I always get the following error running `make` in `PlusBuild-bin/vtk-bin`:

```
Consolidate compiler generated dependencies of target qmlvtkplugin
[ 95%] Linking CXX shared module ../../../lib/qml/VTK.9.1/libqmlvtkpluginD.so
Generating qmltypes file using qmlplugindump
QQmlComponent: Component is not ready
make[2]: *** [GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/build.make:150: lib/qml/VTK.9.1/libqmlvtkpluginD.so] Error 3
make[2]: *** Deleting file 'lib/qml/VTK.9.1/libqmlvtkpluginD.so'
make[1]: *** [CMakeFiles/Makefile2:11476: GUISupport/QtQuick/qml/CMakeFiles/qmlvtkplugin.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Reply by adamrankin** [reply]
Would you be able to delete the vtk-bin, vtk, vtk-int, and vtk-prefix directories, re-run cmake on the superbuild folder, and try a clean build of vtk?

If you've already done that, disregard. Otherwise maybe this discussion might help?
https://gitlab.kitware.com/vtk/vtk/-/issues/18219
